x-pack/plugins/apm/typings/elasticsearch.ts

```diff
@@ -4,24 +4,118 @@
+import { StringMap, IndexAsString } from './common';
 
 declare module 'elasticsearch' {
   // extending SearchResponse to be able to have typed aggregations
-  export interface AggregationSearchResponse<T, U = void>
-    extends SearchResponse<T> {
-    aggregations: U;
-  }
 
-  export interface BucketAgg<T = string> {
-    key: T;
-    doc_count: number;
-  }
+  type AggregationType =
+    | 'date_histogram'
+    | 'histogram'
+    | 'terms'
+    | 'avg'
+    | 'top_hits'
+    | 'max'
+    | 'min'
+    | 'percentiles'
+    | 'sum'
+    | 'extended_stats';
 
-  export interface TermsAggsBucket {
-    key: string;
-    doc_count: number;
-  }
+  type AggOptions = AggregationOptionMap & {
+    [key: string]: any;
+  };
+
+  // eslint-disable-next-line @typescript-eslint/prefer-interface
+  export type AggregationOptionMap = {
+    aggs?: {
+      [aggregationName: string]: {
+        [T in AggregationType]?: AggOptions & AggregationOptionMap
+      };
+    };
+  };
+
+  // eslint-disable-next-line @typescript-eslint/prefer-interface
+  type BucketAggregation<SubAggregationMap, KeyType = string> = {
+    buckets: Array<
+      {
+        key: KeyType;
+        key_as_string: string;
+        doc_count: number;
+      } & (SubAggregationMap extends { aggs: any }
+        ? AggregationResultMap<SubAggregationMap['aggs']>
+        : {})
+    >;
+  };
+
+  type AggregationResultMap<AggregationOption> = IndexAsString<
+    {
+      [AggregationName in keyof AggregationOption]: {
+        avg: {
+          value: number | null;
+        };
+        max: {
+          value: number | null;
+        };
+        min: {
+          value: number | null;
+        };
+        sum: {
+          value: number | null;
+        };
+        terms: BucketAggregation<AggregationOption[AggregationName]>;
+        date_histogram: BucketAggregation<
+          AggregationOption[AggregationName],
+          number
+        >;
+        histogram: BucketAggregation<
+          AggregationOption[AggregationName],
+          number
+        >;
+        top_hits: {
+          hits: {
+            total: number;
+            max_score: number | null;
+            hits: Array<{
+              _source: AggregationOption[AggregationName] extends {
+                Mapping: any;
+              }
+                ? AggregationOption[AggregationName]['Mapping']
+                : never;
+            }>;
+          };
+        };
+        percentiles: {
+          values: {
+            [key: string]: number;
+          };
+        };
+        extended_stats: {
+          count: number;
+          min: number;
+          max: number;
+          avg: number;
+          sum: number;
+          sum_of_squares: number;
+          variance: number;
+          std_deviation: number;
+          std_deviation_bounds: {
+            upper: number;
+            lower: number;
+          };
+        };
+      }[AggregationType & keyof AggregationOption[AggregationName]]
```

Comment:
AggregationOption:
```
{
	name_of_agg: {
		terms: {
			...
		}
	}
}
```

That needs to be translated to:
```
{
	name_of_agg: {
		buckets: []
	}
}
```

That's what this part is for. It picks whatever key from `name_of_agg` that is an aggregation type, and uses that to get the response type from the parent object. Dirty, but works.